presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -118,18 +127,68 @@ public BaseJdbcClient(JdbcConnectorId connectorId, BaseJdbcConfig config, String
         if (config.getConnectionPassword() != null) {
             connectionProperties.setProperty("password", config.getConnectionPassword());
         }
+
+        // generate the map of lowercased schema names to JDBC schema names
+        schemaMappingCache = CacheBuilder.newBuilder().build(new CacheLoader<String, Optional<String>>()
+        {
+            @Override
+            public Optional<String> load(String key)
+                    throws Exception
+            {
+                for (String schemaName : getOriginalSchemas()) {
+                    if (schemaName.equals(key)) {
+                        return Optional.of(schemaName);
+                    }
+
+                    String schemaNameLower = schemaName.toLowerCase(ENGLISH);
+                    if (schemaNameLower.equals(key)) {
+                        return Optional.of(schemaName);
+                    }
+                }
+                return Optional.empty();
+            }
+        });
+
+        if (config.isPreloadSchemaTableMapping()) {
+            reloadCache();
+        }
     }
 
-    @Override
-    public Set<String> getSchemaNames()
+    /**
+     * Invalidates all of the caches and reloads
+     */
+    protected void reloadCache()
+    {
+        schemaMappingCache.invalidateAll();
+
+        // this preloads the list of schema names
+        Set<String> schemas = getSchemaNames();
+
+        // invalidate and remove from the schema table mapping all records, start from scratch
+        for (String key : schemaTableMapping.keySet()) {
+            schemaTableMapping.get(key).invalidateAll();
+            schemaTableMapping.remove(key);
+        }
```

Comment:
any reason you can't just iterate through the values?

    schemaTableMapping.values().foreach(v -> v.invalidateAll());
    schemaTableMapping.clear();

Reply:
I haven't typically used the Java8 Lambdas because of our code base, which sadly has an incredibly out of date Hibernate which barfs if we try and use Lambdas in the wrong place. *sigh*

But yes, that forEach is more efficient, especially the clear() after rather than doing a bunch of removes.